src/zenml/utils/analytics_utils.py

```diff
@@ -565,24 +545,77 @@ def inner_func(*args: Any, **kwargs: Any) -> Any:
         Returns:
             Result of the function.
         """
-        result = func(*args, **kwargs)
-        try:
-            tracker: Optional[AnalyticsTrackerMixin] = None
-            if len(args) and isinstance(args[0], AnalyticsTrackerMixin):
-                tracker = args[0]
-            for obj in [result] + list(args) + list(kwargs.values()):
-                if isinstance(obj, AnalyticsTrackedModelMixin):
-                    obj.track_event(event_name, tracker=tracker)
-                    break
-            else:
-                if tracker:
-                    tracker.track_event(event_name, metadata)
-                else:
-                    track_event(event_name, metadata)
+        with event_handler(event) as handler:
 
-        except Exception as e:
-            logger.debug(f"Analytics tracking failure for {func}: {e}")
+            try:
+                if len(args) and isinstance(args[0], AnalyticsTrackerMixin):
+                    handler.tracker = args[0]
 
-        return result
+                for obj in list(args) + list(kwargs.values()):
+                    if isinstance(obj, AnalyticsTrackedModelMixin):
+                        handler.metadata = obj.get_analytics_metadata()
+                        break
+            except Exception as e:
+                logger.debug(f"Analytics tracking failure for {func}: {e}")
 
+            result = func(*args, **kwargs)
+
+            try:
+                if isinstance(result, AnalyticsTrackedModelMixin):
+                    handler.metadata = result.get_analytics_metadata()
+            except Exception as e:
+                logger.debug(f"Analytics tracking failure for {func}: {e}")
+
+            return result
+
+    inner_func.__doc__ = func.__doc__
     return inner_func
+
+
+class event_handler(object):
+    """Context handler to enable tracking the success status of an event."""
+
+    def __init__(self, event: AnalyticsEvent):
```

Comment:
I thought about this at first. But there are some operations, where the metadata of an event gets generated either during the operation or right at the end of it. In these cases, it would be impossible to feed it through the constructor.

We can still add it though for the example patterns that you have written.

Reply:
Got it. I adjusted it accordingly.